invokeai/backend/model_manager/util/select_hf_files.py

```diff
@@ -0,0 +1,127 @@
+# Copyright (c) 2023 Lincoln D. Stein and the InvokeAI Development Team
+"""
+Select the files from a HuggingFace repository needed for a particular model variant.
+
+Usage:
+```
+from invokeai.backend.model_manager.util.select_hf_files import select_hf_model_files
+from invokeai.backend.model_manager.metadata.fetch import HuggingFaceMetadataFetch
+
+metadata = HuggingFaceMetadataFetch().from_url("https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0")
+files_to_download = select_hf_model_files(metadata.files, variant='onnx')
+```
+"""
+
+import re
+from pathlib import Path
+from typing import Dict, List, Optional, Set
+
+from ..config import ModelRepoVariant
+
+
+def filter_files(
+    files: List[Path],
+    variant: Optional[ModelRepoVariant] = None,
+    subfolder: Optional[Path] = None,
+) -> List[Path]:
+    """
+    Take a list of files in a HuggingFace repo root and return paths to files needed to load the model.
+
+    :param files: List of files relative to the repo root.
+    :param subfolder: Filter by the indicated subfolder.
+    :param variant: Filter by files belonging to a particular variant, such as fp16.
+
+    The file list can be obtained from the `files` field of HuggingFaceMetadata,
+    as defined in `invokeai.backend.model_manager.metadata.metadata_base`.
+    """
+    variant = variant or ModelRepoVariant.DEFAULT
+    paths: List[Path] = []
+
+    # Start by filtering on model file extensions, discarding images, docs, etc
+    for file in files:
+        if file.name.endswith((".json", ".txt")):
+            paths.append(file)
+        elif file.name.endswith(("learned_embeds.bin", "ip_adapter.bin", "lora_weights.safetensors")):
+            paths.append(file)
+        elif re.search(r"model(\.[^.]+)?\.(safetensors|bin|onnx|xml|pth|pt|ckpt|msgpack)$", file.name):
+            paths.append(file)
```

Comment:
Wouldn't we want all model files (`.safetensors`, `.bin`, ... )? Or is that overly permissive for some reason? In other words, why does it have to contain `model.`?

Reply:
The repos often contain checkpoint files as well as the diffusers file, so retrieving all `.safetensors` and so forth sometimes will double the size of the download (or worse). It appears that diffusers components always have `model` somewhere in the name -- at least that's been the case for all the models InvokeAI supports so far. I'll put in a comment to flag this is a potential area of brittleness.